### PR TITLE
Bundle PLY with xonsh

### PR DESCRIPTION
**Pull request by melund**
I haven't added code to use an installed version instead of the bundled. Do we want to do that? In any case, we would have to check the version, and PLY 3.8 doesn't report its correct version number yet (https://github.com/dabeaz/ply/issues/94) ...

### CODE REVIEW

File: .appveyor.yml [hunk 1]
```diff
@@ -1,8 +1,8 @@
 version: 0.2.7.{build}
 os: Windows Server 2012 R2
 install:
-- C:\Python34\Scripts\pip install ply pyreadline nose pygments prompt_toolkit
-- C:\Python35\Scripts\pip install ply pyreadline nose pygments prompt_toolkit
+- C:\Python34\Scripts\pip install pyreadline nose pygments prompt_toolkit
+- C:\Python35\Scripts\pip install pyreadline nose pygments prompt_toolkit
```
Comment (scopatz) [inline .appveyor.yml]: We probably should still install ply so that we are testing the latest version on pypi

File: docs/dependencies.rst [hunk 2]
```diff
@@ -5,7 +5,6 @@ Xonsh currently has the following external dependencies,
 *Run Time:*
 
     #. Python v3.4+
-    #. PLY
```
Comment (scopatz) [inline docs/dependencies.rst]: I would change this to `#. PLY (optional, included with xonsh)` or similar

File: docs/index.rst [hunk 3]
```diff
@@ -221,7 +221,6 @@ Xonsh currently has the following external dependencies,
 *Run Time:*
 
     #. Python v3.4+
-    #. PLY
```
Comment (scopatz) [inline docs/index.rst]: Same here

File: recipe/meta.yaml [hunk 4]
```diff
@@ -15,12 +15,10 @@ build:
 requirements:
   build:
     - python
-    - ply
```
Comment (scopatz) [inline recipe/meta.yaml]: Also, I would keep these around so that the dependency is managed elsewhere.

File: tests/test_lexer.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 
 import nose
 
-from ply.lex import LexToken
+from xonsh.ply.lex import LexToken
```
Comment (scopatz) [inline tests/test_lexer.py]: I would change this to 

```python
try:
    from ply.lex import LexToken
except ImportError:
    from xonsh.ply.lex import LexToken
```

File: xonsh/lexer.py [hunk 6]
```diff
@@ -8,7 +8,7 @@
 from io import BytesIO
 from keyword import kwlist
 
-from ply.lex import LexToken
+from xonsh.ply.lex import LexToken
```
Comment (scopatz) [inline xonsh/lexer.py]: Same

File: xonsh/parsers/base.py [hunk 7]
```diff
@@ -1,10 +1,9 @@
 # -*- coding: utf-8 -*-
 """Implements the base xonsh parser."""
 from collections import Iterable, Sequence, Mapping
-
-from ply import yacc
```
Comment (scopatz) [inline xonsh/parsers/base.py]: Same try-except idea here.

File: xonsh/xonfig.py [hunk 8]
```diff
@@ -9,8 +9,7 @@
 from pprint import pformat
 from argparse import ArgumentParser
 
-import ply
```
Comment (scopatz) [inline xonsh/xonfig.py]: Same try-except idea here.

**Comment by scopatz** [comment]
Thanks a ton for putting this in @melund! I haven't gotten around to it myself :)

**Comment by melund** [comment]
I also did the try-import blocks first. But then I was wondering if needed to add test so we don't use older installed versions PLY. Then I discovered that 3.8 was reporting it self as 3.7 (a bug) and decided to just do the simple thing first.

@scopatz if you think we don't need to check the version then I will add `try` blocks back.

**Comment by scopatz** [comment]
@melund - yeah, I don't think we need to test the version.  It shouldn't matter if the user wants to have an older version of PLY installed.

**Comment by scopatz** [comment]
At least IMHO

**Comment by scopatz** [comment]
Is this ready for another look @melund, or should we hold off?

**Comment by melund** [comment]
@scopatz I think is good to merge. I will think about the subtree stuff, but I think it can wait.

**Comment by scopatz** [comment]
@melund woohoo! Thanks!

**Pull request merged**